hook-janitor/src/webhooks.rs

```diff
@@ -1,55 +1,454 @@
+                Ok(future) => delivery_futures.push(future),
+                Err((error, _)) => return Err(WebhookCleanerError::KafkaProduceError { error }),
+            }
+        }
+
+        for result in join_all(delivery_futures).await {
+            match result {
+                Ok(Ok(_)) => {}
+                Ok(Err((error, _))) => {
+                    return Err(WebhookCleanerError::KafkaProduceError { error })
+                }
+                Err(_) => {
+                    // Cancelled due to timeout while retrying
+                    return Err(WebhookCleanerError::KafkaProduceCanceled);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn delete_observed_rows(&self, tx: &mut Transaction<'_, Postgres>) -> Result<u64> {
+        // This DELETE is only safe because we are in serializable isolation mode, see the note
+        // in `start_serializable_txn`.
+        let base_query = format!(
+            r#"
+            DELETE FROM {0}
+            WHERE status IN ('failed', 'completed')
+              AND queue = $1;
+            "#,
+            self.table_name
+        );
+
+        let result = sqlx::query(&base_query)
+            .bind(&self.queue_name)
+            .execute(&mut **tx)
+            .await
+            .map_err(|e| WebhookCleanerError::DeleteRowsError { error: e })?;
+
+        Ok(result.rows_affected())
+    }
+
+    async fn commit_txn(&self, tx: Transaction<'_, Postgres>) -> Result<()> {
+        tx.commit()
+            .await
+            .map_err(|e| WebhookCleanerError::CommitTxnError { error: e })?;
+
+        Ok(())
+    }
+
+    async fn cleanup_impl(&self) -> Result<()> {
+        debug!("WebhookCleaner starting cleanup");
+
+        // Note that we select all completed and failed rows without any pagination at the moment.
+        // We aggregrate as much as possible with GROUP BY, truncating the timestamp down to the
+        // hour just like App Metrics does. A completed row is 24 bytes (and aggregates an entire
+        // hour per `plugin_config_id`), and a failed row is 104 bytes + the error message length
+        // (and aggregates an entire hour per `plugin_config_id` per `error`), so we can fit a lot
+        // of rows in memory. It seems unlikely we'll need to paginate, but that can be added in the
+        // future if necessary.
+
+        let mut tx = self.start_serializable_txn().await?;
+
+        let completed_rows = self.get_completed_rows(&mut tx).await?;
+        let completed_agg_row_count = completed_rows.len();
+        let completed_kafka_payloads = self.serialize_completed_rows(completed_rows)?;
+
+        let failed_rows = self.get_failed_rows(&mut tx).await?;
+        let failed_agg_row_count = failed_rows.len();
+        let mut failed_kafka_payloads = self.serialize_failed_rows(failed_rows)?;
+
+        let mut all_kafka_payloads = completed_kafka_payloads;
+        all_kafka_payloads.append(&mut failed_kafka_payloads);
```

Comment:
I'd prefer we use `extend` here. `extend` consumes the `failed_kafka_payloads` vec, so we won't be able to access it later. By using append, we are still keeping ownership of `failed_kafka_payloads`, opening the door up for bugs. Not seeing any bugs right now, but just feel like it's safer.

Reply:
Ah, I missed that we weren't giving up ownership with `append`, thanks.